File: Java-Core/src/main/java/presto/grammar/ItemInstance.java
package presto.grammar;

import presto.error.IndexOutOfRangeError;
import presto.error.InvalidDataError;
import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.expression.IExpression;
import presto.runtime.Context;
import presto.type.IType;
import presto.type.IntegerType;
import presto.utils.CodeWriter;
import presto.value.ICollection;
import presto.value.IValue;
import presto.value.Integer;
import presto.value.ListValue;

public class ItemInstance implements IAssignableSelector {

	IAssignableInstance parent;
	IExpression item;
	
	public ItemInstance(IExpression item) {
		this.item = item;
	}
	
	@Override
	public void setParent(IAssignableInstance parent) {
		this.parent = parent;
	}
	
	public IExpression getItem() {
		return item;
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		parent.toDialect(writer);
		writer.append('[');
		item.toDialect(writer);
		writer.append(']');
	}
	
	@Override
	public void checkAssignValue(Context context, IExpression expression) throws SyntaxError {
		parent.checkAssignElement(context);
		IType itemType = item.check(context);
		if(itemType!=IntegerType.instance())
			throw new SyntaxError("Expecting an Integer, got:" + itemType.toString());
	}
	
	@Override
	public void checkAssignMember(Context context, String memberName) throws SyntaxError {
		// TODO Auto-generated method stub
		
	}
	
	@Override
	public void checkAssignElement(Context context) throws SyntaxError {
		// TODO Auto-generated method stub
		
	}
	
	@Override
	public void assign(Context context, IExpression expression) throws PrestoError {
		IValue obj = parent.interpret(context);
		if(!(obj instanceof ListValue))
			throw new InvalidDataError("Expected a List, got:" + obj.getClass().getName());
		ListValue list = (ListValue)obj;
		IValue idx = item.interpret(context);
		if(!(idx instanceof Integer))
			throw new InvalidDataError("Expected an Integer, got:" + idx.getClass().getName());
		int index = (int)((Integer)idx).IntegerValue();
		if(index<1 || index>list.length())
			throw new IndexOutOfRangeError();
		list.setItem(index-1, expression.interpret(context));
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		IValue parent = this.parent.interpret(context);
		IValue item = this.item.interpret(context);
        if (parent instanceof ICollection)
            return ((ICollection<?>)parent).getItem(context, item);
         else
 			throw new SyntaxError("Unknown item/key: " + item.getClass().getName());
	}
	

}

File: Java-CodeServer/src/main/resources/js/prompto/grammar/ItemInstance.js
var InvalidDataError = require("../error/InvalidDataError").InvalidDataError;
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var IntegerType = require("../type/IntegerType").IntegerType;
var BaseValueList = require("../value/BaseValueList").BaseValueList;
var Integer = require("../value/Integer").Integer;
var Value = require("../value/Value").Value;

function ItemInstance(item) {
	this.parent = null;
	this.item = item;
	return this;
}
	
ItemInstance.prototype.toString = function() {
	return this.parent.toString() + "[" + this.item.toString() + "]";
};

ItemInstance.prototype.toDialect = function(writer) {
    this.parent.toDialect(writer);
    writer.append('[');
    this.item.toDialect(writer);
    writer.append(']');
}

ItemInstance.prototype.checkAssignValue = function(context, expression) {
	this.parent.checkAssignElement(context);
	var itemType = this.item.check(context);
	if(itemType!=IntegerType.instance) {
		throw new SyntaxError("Expecting an Integer, got:" + itemType.toString());
	}
};

ItemInstance.prototype.checkAssignMember = function(context, memberName) {
	// TODO Auto-generated method stub
};

/*

@Override
public void checkAssignElement(Context context) throws SyntaxError {
	// TODO Auto-generated method stub
	
}

*/

ItemInstance.prototype.assign = function(context, expression) {
	var obj = this.parent.interpret(context);
	if(!(obj instanceof BaseValueList)) {
		throw new InvalidDataError("Expected a List, got:" + typeof(obj));
	}
	var idx = this.item.interpret(context);
	if(!(idx instanceof Integer)) {
		throw new InvalidDataError("Expected an Integer, got:" + idx.getClass().getName());
	}
	var index = idx.IntegerValue();
	if(index<1 || index>obj.size()) {
		throw new IndexOutOfRangeError();
	}
	obj.setItem(index-1, expression.interpret(context));
};

ItemInstance.prototype.interpret = function(context) {
    var parent = this.parent.interpret(context);
    var item = this.item.interpret(context);
    if (parent.getItemInContext && item instanceof Value) {
		return parent.getItemInContext(context, item);
	} else {
		throw new SyntaxError("Unknown item/key: " + typeof(item));
	}
};

exports.ItemInstance = ItemInstance;
